### PR TITLE
Automatically close the registration when the admin is inactive

### PR DESCRIPTION
**Pull request by annando**
This PR introduces a mechanism to automatically close the registrations when an admin isn't active any more on the system.

### CODE REVIEW

**Comment by AlfredSK** [comment]
> automatically close the registrations when **an** admin isn't active any more on the system.

Just a question. Do you literally mean **an** admin or **all** admins?

**Comment by annando** [comment]
There need to be at least one admin who is active. Means that when you have got three admin accounts, only one need to be active in a range of 30 days. And with "active" it means that the person has to use Friendica via the web interface in either their own user account or one of their delegated accounts. It isn't enough to use Friendica via the API, since you could have installed some app on your mobile device that checks for notifications but you never really use it.

File: src/Content/Post/Entity/PostMedia.php [hunk 1]
```diff
@@ -302,4 +302,19 @@ public function hasDimensions(): bool
 	{
 		return $this->width && $this->height || $this->previewWidth && $this->previewHeight;
 	}
+
+	public function copyImageToPreview()
```
Comment (MrPetovan) [inline src/Content/Post/Entity/PostMedia.php]: This shouldn't be part of this PR.

File: src/Module/Api/GNUSocial/GNUSocial/Config.php [hunk 2]
```diff
@@ -47,7 +48,7 @@ protected function rawContent(array $request = [])
 				'broughtby'    => '',
 				'broughtbyurl' => '',
 				'timezone'     => DI::config()->get('system', 'default_timezone'),
-				'closed'       => (DI::config()->get('config', 'register_policy') == Register::CLOSED),
+				'closed'       => (Security::getRegisterPolicy() == Register::CLOSED),
```
Comment (MrPetovan) [inline src/Module/Api/GNUSocial/GNUSocial/Config.php]: After changing the return type of `getRegisterPolicy`
```suggestion
				'closed'       => Security::getRegisterPolicy() === Register::CLOSED,
```

File: src/Module/Bookmarklet.php [hunk 3]
```diff
@@ -42,7 +43,7 @@ protected function content(array $request = []): string
 
 		if (!DI::userSession()->getLocalUserId()) {
 			$output = '<h2>' . DI::l10n()->t('Login') . '</h2>';
-			$output .= Login::form(DI::args()->getQueryString(), intval($config->get('config', 'register_policy')) === Register::CLOSED ? false : true);
+			$output .= Login::form(DI::args()->getQueryString(), Security::getRegisterPolicy() === Register::CLOSED ? false : true);
```
Comment (MrPetovan) [inline src/Module/Bookmarklet.php]:
```suggestion
			$output .= Login::form(DI::args()->getQueryString(), Security::getRegisterPolicy() !== Register::CLOSED);
```

File: src/Module/Home.php [hunk 4]
```diff
@@ -73,7 +74,7 @@ protected function content(array $request = []): string
 			}
 		}
 
-		$login = Login::form(DI::args()->getQueryString(), $config->get('config', 'register_policy') === Register::CLOSED ? 0 : 1);
+		$login = Login::form(DI::args()->getQueryString(), Security::getRegisterPolicy() === Register::CLOSED ? 0 : 1);
```
Comment (MrPetovan) [inline src/Module/Home.php]:
```suggestion
		$login = Login::form(DI::args()->getQueryString(), Security::getRegisterPolicy() !== Register::CLOSED);
```

File: src/Object/Api/Mastodon/Instance.php [hunk 5]
```diff
@@ -75,7 +76,7 @@ class Instance extends BaseDataTransferObject
 
 	public function __construct(IManageConfigValues $config, BaseURL $baseUrl, Database $database, Configuration $configuration, ?Account $contact_account, array $rules)
 	{
-		$register_policy = intval($config->get('config', 'register_policy'));
+		$register_policy = Security::getRegisterPolicy();
```
Comment (MrPetovan) [inline src/Object/Api/Mastodon/Instance.php]: The operators need to be stricter on lines 91-92.

File: src/Module/Api/Mastodon/InstanceV2.php [hunk 6]
```diff
@@ -166,7 +167,7 @@ private function buildFriendicaExtensionInfo(): InstanceEntity\FriendicaExtensio
 
 	private function buildRegistrationsInfo(): InstanceEntity\Registrations
 	{
-		$register_policy   = intval($this->config->get('config', 'register_policy'));
+		$register_policy   = Security::getRegisterPolicy();
 		$enabled           = ($register_policy != Register::CLOSED);
 		$approval_required = ($register_policy == Register::APPROVE);
```
Comment (MrPetovan) [inline src/Module/Api/Mastodon/InstanceV2.php]:
```suggestion
		$enabled           = $register_policy !== Register::CLOSED;
		$approval_required = $register_policy === Register::APPROVE;
```

File: src/Content/Nav.php [hunk 7]
```diff
@@ -251,7 +252,7 @@ private function getInfo(): array
 			$nav['home'] = [$homelink, $this->l10n->t('Home'), '', $this->l10n->t('Home Page')];
 		}
 
-		if (intval($this->config->get('config', 'register_policy')) === \Friendica\Module\Register::OPEN && !$this->session->isAuthenticated()) {
+		if (intval(Security::getRegisterPolicy()) === \Friendica\Module\Register::OPEN && !$this->session->isAuthenticated()) {
```
Comment (MrPetovan) [inline src/Content/Nav.php]:
```suggestion
		if (Security::getRegisterPolicy() === \Friendica\Module\Register::OPEN && !$this->session->isAuthenticated()) {
```

File: src/Module/Register.php [hunk 8]
```diff
@@ -86,7 +87,7 @@ protected function content(array $request = []): string
 			}
 		}
 
-		if (!DI::userSession()->getLocalUserId() && (intval(DI::config()->get('config', 'register_policy')) === self::CLOSED)) {
+		if (!DI::userSession()->getLocalUserId() && (Security::getRegisterPolicy() === self::CLOSED)) {
```
Comment (MrPetovan) [inline src/Module/Register.php]: Save the parentheses!
```suggestion
		if (!DI::userSession()->getLocalUserId() && Security::getRegisterPolicy() === self::CLOSED) {
```

File: src/Module/Security/Login.php [hunk 9]
```diff
@@ -118,7 +119,7 @@ public static function form(string $return_path = null, bool $register = false):
 		}
 
 		$reg = false;
-		if ($register && intval(DI::config()->get('config', 'register_policy')) !== Register::CLOSED) {
+		if ($register && Security::getRegisterPolicy() !== Register::CLOSED) {
```
Comment (MrPetovan) [inline src/Module/Security/Login.php]: The `$register` parameter is set across modules using `Security::getRegisterPolicy() !== Register::CLOSED`, so it feels redundant.

If the default is to show the register link except if the register disallow it or the parameter is manually set to false, the default value should be `true` and several calls to `Login::form()` can be simplified.

File: src/Module/User/Import.php [hunk 10]
```diff
@@ -80,7 +81,7 @@ public function __construct(PermissionSet $permissionSet, IManagePersonalConfigV
 
 	protected function post(array $request = [])
 	{
-		if ($this->config->get('config', 'register_policy') != \Friendica\Module\Register::OPEN && !$this->app->isSiteAdmin()) {
+		if (Security::getRegisterPolicy() != \Friendica\Module\Register::OPEN && !$this->app->isSiteAdmin()) {
```
Comment (MrPetovan) [inline src/Module/User/Import.php]:
```suggestion
		if (Security::getRegisterPolicy() !== \Friendica\Module\Register::OPEN && !$this->app->isSiteAdmin()) {
```

File: src/Module/User/Import.php [hunk 11]
```diff
@@ -99,7 +100,7 @@ protected function post(array $request = [])
 
 	protected function content(array $request = []): string
 	{
-		if (($this->config->get('config', 'register_policy') != \Friendica\Module\Register::OPEN) && !$this->app->isSiteAdmin()) {
+		if ((Security::getRegisterPolicy() != \Friendica\Module\Register::OPEN) && !$this->app->isSiteAdmin()) {
```
Comment (MrPetovan) [inline src/Module/User/Import.php]:
```suggestion
		if ((Security::getRegisterPolicy() !== \Friendica\Module\Register::OPEN) && !$this->app->isSiteAdmin()) {
```

File: src/Object/Api/Mastodon/Instance.php (hunk 5 above)
Comment (MrPetovan) [inline src/Object/Api/Mastodon/Instance.php]: Operators need to be tightened on lines 91-92.

File: src/Security/Security.php [hunk 12]
```diff
@@ -138,4 +140,20 @@ public static function getPermissionsSQLByUserId(int $owner_id, bool $accessible
 		}
 		return $sql;
 	}
+
+	public static function getRegisterPolicy(): int
```
Comment (MrPetovan) [inline src/Security/Security.php]: Since the registration policy constants are defined in the `Module\Register` class, I believe this method should be moved over there as well. Or the constants be moved in the `Security` class, but it doesn't feel right to me. I'll approve either way though.

File: src/Security/Security.php [hunk 13]
```diff
@@ -138,4 +140,20 @@ public static function getPermissionsSQLByUserId(int $owner_id, bool $accessible
...
+	{
+		$days = DI::config()->get('system', 'admin_inactivity_limit');
+		if ($days == 0) {
+			return DI::config()->get('config', 'register_policy');
```
Comment (MrPetovan) [inline src/Security/Security.php]:
```suggestion
			return intval(DI::config()->get('config', 'register_policy'));
```

File: src/Security/Security.php [hunk 14]
```diff
@@ -138,4 +140,20 @@ public static function getPermissionsSQLByUserId(int $owner_id, bool $accessible
...
+		}
+
+		$inactive_since = DateTimeFormat::utc('now - ' . $days . ' day');
+		foreach (User::getAdminList(['login_date']) as $admin) {
+			if (strtotime($admin['login_date']) > strtotime($inactive_since)) {
+				return DI::config()->get('config', 'register_policy');
```
Comment (MrPetovan) [inline src/Security/Security.php]:
```suggestion
				return intval(DI::config()->get('config', 'register_policy'));
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Content/Post/Entity/PostMedia.php (hunk 1 above)
Comment (annando) [reply src/Content/Post/Entity/PostMedia.php]: Could be that something slipped through here. I worked on a ton of branches in parallel. I will check later.

File: src/Security/Security.php (hunk 12 above)
Comment (annando) [reply src/Security/Security.php]: I wasn't totally sure about the right place either.

File: src/Object/Api/Mastodon/Instance.php (hunk 5 above)
Comment (annando) [reply src/Object/Api/Mastodon/Instance.php]: Done

Comment (annando) [reply src/Object/Api/Mastodon/Instance.php]: Done

File: src/Security/Security.php (hunk 12 above)
Comment (MrPetovan) [reply src/Security/Security.php]: Like I said, it doesn't really matter as long as both the constants and this method are in the same class. Since we use these constants in the config files, it'd be less disruptive to move this method in the Register module.

Comment (annando) [reply src/Security/Security.php]: Currently the register class hasn't a `DI` implementation, also moving this would mean to possibly enhance a handful of `initialize` calls as well. Since I was slightly lazy, I wanted to avoid this 😁

Comment (MrPetovan) [reply src/Security/Security.php]: I didn't expect you to make it a class method, you can move it as a static method because it doesn't need all the module dependencies. If you don't want to add more `DI` calls to the `Register` module class, you can make the `Module\Register::getPolicy` (after move/rename) method accept a `IManageConfigValues` parameter.

Comment (annando) [reply src/Security/Security.php]: Isn't it unwanted to have static functions inside of classes where all the other ones aren't? if this is okay, I can do it that way. I just wanted to avoid too many changes, since I had to change a lot of places there.

Comment (MrPetovan) [reply src/Security/Security.php]: No, it's fine because the method is outside of the regular lifecycle of Module objects of processing a HTTP request into a HTTP response. As a result it has few dependencies (except `User` but we aren't fixing this any time soon) and it isn't a problem for it to be static.

File: src/Model/Item.php [hunk 15]
```diff
@@ -3410,14 +3410,14 @@ public static function prepareBody(array &$item, bool $attach = false, bool $is_
 		if (!empty($shared_item['uri-id'])) {
 			$shared_uri_id = $shared_item['uri-id'];
 			$shared_links[] = strtolower($shared_item['plink']);
-			$sharedSplitAttachments = DI::postMediaRepository()->splitAttachments($shared_uri_id, [], $shared_item['has-media']);
+			$sharedSplitAttachments = DI::postMediaRepository()->splitAttachments($shared_uri_id, [], $shared_item['has-media'], $sensitive);
```
Comment (MrPetovan) [inline src/Model/Item.php]: This shouldn't be part of this PR either.

File: src/Module/Register.php [hunk 16]
```diff
@@ -426,4 +427,19 @@ private function sendNotification(array $user, string $event)
 			]);
 		}
 	}
+	public static function getRegisterPolicy(): int
```
Comment (MrPetovan) [inline src/Module/Register.php]: Class is already called `Register`.
```suggestion
	public static function getPolicy(): int
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Module/Register.php (hunk 16 above)
Comment (annando) [reply src/Module/Register.php]: Done

**Review by MrPetovan — APPROVED** [review]
